tirocks/src/option/read.rs

```diff
@@ -0,0 +1,247 @@
+// Copyright 2022 TiKV Project Authors. Licensed under Apache-2.0.
+
+use std::{
+    fmt::{self, Debug, Formatter},
+    mem::MaybeUninit,
+};
+
+use tirocks_sys::rocksdb_titandb_TitanReadOptions;
+
+use super::OwnedSlie;
+
+pub type ReadTier = tirocks_sys::rocksdb_ReadTier;
+pub type SequenceNumber = tirocks_sys::rocksdb_SequenceNumber;
+
+/// Options that control read operations
+pub struct ReadOptions {
+    raw: rocksdb_titandb_TitanReadOptions,
+    // Storage for iterate_lower_bound, iterate_upper_bound and timestamp.
+    slice_store: Option<Box<[OwnedSlie; 3]>>,
+}
+
+impl Default for ReadOptions {
+    #[inline]
+    fn default() -> ReadOptions {
+        let mut opt = MaybeUninit::uninit();
+        unsafe {
+            tirocks_sys::ctitandb_readoptions_init(opt.as_mut_ptr());
+            ReadOptions {
+                raw: opt.assume_init(),
+                slice_store: None,
+            }
+        }
+    }
+}
+
+impl Debug for ReadOptions {
+    fn fmt(&self, f: &mut Formatter<'_>) -> fmt::Result {
+        write!(f, "{:?}", self.raw)
+    }
+}
+
+impl ReadOptions {
+    #[inline]
+    fn init_slice_store(&mut self) {
+        if self.slice_store.is_none() {
+            self.slice_store = Some(Box::new(Default::default()));
+        }
+    }
+
+    /// Sets the smallest key at which the backward iterator can return an
+    /// entry. Once the bound is passed, Valid() will be false. It is inclusive
+    /// ie the bound value is a valid entry.
+    ///
+    /// If prefix_extractor is set, the Seek target and `iterate_lower_bound`
+    /// need to have the same prefix. This is because ordering is not guaranteed
+    /// outside of prefix domain.
+    #[inline]
+    pub fn set_iterate_lower_bound(&mut self, lower_bound: Vec<u8>) -> &mut Self {
+        self.init_slice_store();
+        let arr = self.slice_store.as_mut().unwrap();
+        arr[0].set_data(lower_bound);
+        self.raw._base.iterate_lower_bound = &mut arr[0].slice;
```

Comment:
Is this a standard way to access base class?

Reply:
From C++ perspective, no. The standard way should use `.` or `->` instead of referencing its base class and let compiler to compute the correct offset. In Rust bindings also no, as there is no "standard". Under the hook, C++ will put the member fields of base class first and then the derived class. As long as the bindings matches the class/struct memory layout, it's OK to access the base class explicitly because it compiles to address in the end.